source/common/upstream/upstream_impl.cc

```diff
@@ -434,14 +414,52 @@ ClusterSharedPtr ClusterImplBase::create(const envoy::api::v2::Cluster& cluster,
 }
 
+Stats::ScopePtr ClusterImplBase::generateStatsScope(const envoy::api::v2::Cluster& config,
+                                                    Stats::Store& stats) {
+  return stats.createScope(fmt::format("cluster.{}.", config.alt_stat_name().empty()
+                                                          ? config.name()
+                                                          : std::string(config.alt_stat_name())));
+}
+
+Network::TransportSocketFactoryPtr ClusterImplBase::createTransportSocketFactory(
+    const envoy::api::v2::Cluster& config, Stats::Scope& stats_scope,
+    Ssl::ContextManager& ssl_context_manager, Secret::SecretManager& secret_manager,
+    Init::Manager& init_manager) {
+  // If the cluster config doesn't have a transport socket configured, override with the default
+  // transport socket implementation based on the tls_context. We copy by value first then override
+  // if necessary.
+  auto transport_socket = config.transport_socket();
+  if (!config.has_transport_socket()) {
+    if (config.has_tls_context()) {
+      transport_socket.set_name(Extensions::TransportSockets::TransportSocketNames::get().TLS);
+      MessageUtil::jsonConvert(config.tls_context(), *transport_socket.mutable_config());
+    } else {
+      transport_socket.set_name(
+          Extensions::TransportSockets::TransportSocketNames::get().RAW_BUFFER);
+    }
+  }
+
+  Server::Configuration::TransportSocketFactoryContextImpl factory_context(
+      ssl_context_manager, stats_scope, secret_manager, init_manager);
+  auto& config_factory = Config::Utility::getAndCheckFactory<
+      Server::Configuration::UpstreamTransportSocketConfigFactory>(transport_socket.name());
+  ProtobufTypes::MessagePtr message =
+      Config::Utility::translateToFactoryConfig(transport_socket, config_factory);
+
+  return config_factory.createTransportSocketFactory(*message, factory_context);
+}
+
 ClusterImplBase::ClusterImplBase(const envoy::api::v2::Cluster& cluster,
                                  const envoy::api::v2::core::BindConfig& bind_config,
                                  Runtime::Loader& runtime, Stats::Store& stats,
                                  Ssl::ContextManager& ssl_context_manager,
                                  Secret::SecretManager& secret_manager, bool added_via_api)
-    : runtime_(runtime),
-      info_(new ClusterInfoImpl(cluster, bind_config, runtime, stats, ssl_context_manager,
-                                secret_manager, added_via_api)) {
+    : runtime_(runtime), stats_scope_(generateStatsScope(cluster, stats)),
+      info_(new ClusterInfoImpl(cluster, bind_config, runtime,
+                                createTransportSocketFactory(cluster, *stats_scope_.get(),
+                                                             ssl_context_manager, secret_manager,
+                                                             init_manager_),
+                                stats_scope_.release(), added_via_api)) {
```

Comment:
You need following in the  ClusterImplBase constuctor

 ClusterImplBase:: ClusterImplBase(...) {
    auto stat_scope = generateStatsScope(cluster, stats);
    auto socket_factory = createTransportSocketFactory(cluster, *stat_scope, ...);
    info_ = std::make_unique<ClusterInfoImpl>(...,  socket_factory,  std::move(stat_scope), ...);
...
}